src/sentry/sentry_metrics/multiprocess.py

```diff
@@ -0,0 +1,406 @@
+        assert not self.__closed
+
+        if len(self.__futures) >= self.__max_buffer_size:
+            raise MessageRejected
+
+        for message in outer_message.payload:
+            payload = message.payload
+            future = self.__producer.produce(
+                destination=Topic(self.__producer_topic),
+                payload=payload,
+            )
+            self.__futures.append(ProducerResultFuture(message, future))
+
+    def close(self) -> None:
+        self.__closed = True
+
+    def terminate(self) -> None:
+        self.__closed = True
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        start = time.time()
+        while self.__futures:
+            remaining = timeout - (time.time() - start) if timeout is not None else None
+            if remaining is not None and remaining <= 0:
+                logger.warning(f"Timed out with {len(self.__futures)} futures in queue")
+                break
+
+            message, future = self.__futures.popleft()
+
+            future.result(remaining)
+
+            self.__commit_function({message.partition: Position(message.offset, message.timestamp)})
+
+
+def process_messages(
+    outer_message: Message[MessageBatch],
+) -> MessageBatch:
+    """
+    We have an outer_message Message() whose payload is a batch of Message() objects.
+
+        Message(
+            partition=...,
+            offset=...
+            timestamp=...
+            payload=[Message(...), Message(...), etc]
+        )
+
+    The inner messages payloads are KafkaPayload's that have:
+        * key
+        * headers
+        * value
+
+    The value of the message is what we need to parse and then translate
+    using the indexer.
+    """
+    indexer = get_indexer()
+    metrics = get_metrics()
+    task = get_task()
+
+    strings = set()
+    parsed_payloads_by_offset = {
+        msg.offset: json.loads(msg.payload.value.decode("utf-8"), use_rapid_json=True)
+        for msg in outer_message.payload
+    }
+
+    for message in parsed_payloads_by_offset.values():
+        metric_name = message["name"]
+        tags = message.get("tags", {})
+
+        parsed_strings = {
+            metric_name,
+            *tags.keys(),
+            *tags.values(),
+        }
+        strings.update(parsed_strings)
+
+    with metrics.timer("metrics_consumer.bulk_record"):
+        mapping = indexer.bulk_record(list(strings))  # type: ignore
+
+    new_messages: Sequence[Message[KafkaPayload]] = []
+    celery_messages: Sequence[Mapping[str, Union[str, int, Mapping[int, int]]]] = []
+
+    for message in outer_message.payload:
+        parsed_payload_value = parsed_payloads_by_offset[message.offset]
+        new_payload_value = parsed_payload_value
+
+        message_type = parsed_payload_value.get("type", "unknown")
+        metrics.incr(
+            "metrics_consumer.process_message.messages_seen", tags={"metric_type": message_type}
+        )
```

Comment:
You can record this metric once per outer message instead of once per inner message, that would add a lot less overhead.

Reply:
@fpacifici ok i can scrap the metric type, not sure that info is actually necessary here